folium/features.py

```diff
@@ -519,10 +570,77 @@ def _get_self_bounds(self):
 
     def render(self, **kwargs):
         self.parent_map = get_obj_in_upper_tree(self, Map)
-        self.json = self.style_data() if self.embed else json.dumps(self.data)
+        if self.style or self.highlight:
+            mapper = GeoJsonStyleMapper(self.data, self.find_identifier(),
+                                        self)
+            if self.style:
+                self.style_map = mapper.get_style_map(self.style_function)
+            if self.highlight:
+                self.highlight_map = mapper.get_highlight_map(
+                    self.highlight_function)
+        if self.embed:
+            self.json = json.dumps(self.data, sort_keys=True)
         super(GeoJson, self).render()
 
 
+class GeoJsonStyleMapper:
+    """Create dicts that map styling to GeoJson features.
+
+    Used in the GeoJson class. Users don't have to call this class directly.
+    """
+
+    def __init__(self, data, feature_identifier, geojson_obj):
+        self.data = data
+        self.feature_identifier = feature_identifier
+        self.geojson_obj = geojson_obj
+
+    def get_style_map(self, style_function):
+        """Return a dict that maps style parameters to features."""
+        return self._create_mapping(style_function, 'style')
+
+    def get_highlight_map(self, highlight_function):
+        """Return a dict that maps highlight parameters to features."""
+        return self._create_mapping(highlight_function, 'highlight')
+
+    def _create_mapping(self, func, switch):
+        """Internal function to create the mapping."""
+        mapping = {}
+        for feature in self.data['features']:
+            content = func(feature)
+            if switch == 'style':
+                for key, value in content.items():
+                    if isinstance(value, MacroElement):
+                        # Make sure objects are rendered:
+                        if value._parent is None:
+                            value._parent = self.geojson_obj
+                            value.render()
+                        # Replace objects with their Javascript var names:
+                        content[key] = "{{'" + value.get_name() + "'}}"
+            key = self._to_key(content)
+            mapping.setdefault(key, []).append(self.get_feature_id(feature))
+        self._set_default_key(mapping)
+        return mapping
+
+    def get_feature_id(self, feature):
+        """Return a value identifying the feature."""
+        fields = self.feature_identifier.split('.')[1:]
+        return functools.reduce(operator.getitem, fields, feature)
```

Comment:
Nice!

Reply:
Thanks :) Found it on stackoverflow, looks a bit like magic. I'll definitely write a test for it.